### ISSUE TITLE
Cannot embed story in Docs MDX due to Angular component getting declared multiple times

### ISSUE THREAD

**Issue by ssams** [storybookjs/storybook#17269]
**Describe the bug**

When embedding stories in a documentation page, switching between individual story pages and the docs page triggers errors on the console, and content fails to load on the documentation page:

```
Unhandled Promise rejection: Type ButtonComponent is part of the declarations of 2 modules: StorybookModule and StorybookModule! Please consider moving ButtonComponent to a higher module that imports StorybookModule and StorybookModule. You can also create a new NgModule that exports and includes ButtonComponent then import that NgModule in StorybookModule and StorybookModule.
```

Additional errors are present in the log immediately before this, not sure if they are significant or just a side effect of the other error:
```
Warning: render(...): It looks like the React-rendered content of this container was removed without using React. This is not supported and will cause errors. Instead, call ReactDOM.unmountComponentAtNode to empty a container.

ERROR Error: The selector "example-button--primary-jilgqplasf" did not match any elements

Unhandled Promise rejection: The selector "example-button--primary-jilgqplasf" did not match any elements
```

**To Reproduce**

https://github.com/ssams/storybook-angular-redeclared-comp

Setup with `npx sb@next repro`, adds a single documentation page (`src/stories/Button.stories.mdx`).

To trigger the error, open a single story page of the button component (URL path `/?path=/story/example-button--primary`), then navigate to the added docs page (`/?path=/story/documentation-button--page`).

**System**
```
Environment Info:

  System:
    OS: Linux 5.15 Fedora Linux 35 (Workstation Edition)
    CPU: (8) x64 Intel(R) Core(TM) i7-4790 CPU @ 3.60GHz
  Binaries:
    Node: 14.18.2 - /usr/bin/node
    Yarn: 3.1.1 - /usr/local/bin/yarn
    npm: 6.14.15 - /usr/bin/npm
  Browsers:
    Firefox: 96.0
  npmPackages:
    @storybook/addon-actions: ^6.5.0-alpha.19 => 6.5.0-alpha.19 
    @storybook/addon-docs: ^6.5.0-alpha.19 => 6.5.0-alpha.19 
    @storybook/addon-essentials: ^6.5.0-alpha.19 => 6.5.0-alpha.19 
    @storybook/addon-links: ^6.5.0-alpha.19 => 6.5.0-alpha.19 
    @storybook/angular: ^6.5.0-alpha.19 => 6.5.0-alpha.19 
    @storybook/builder-webpack5: ^6.5.0-alpha.19 => 6.5.0-alpha.19 
    @storybook/manager-webpack5: ^6.5.0-alpha.19 => 6.5.0-alpha.19 
```

**Additional context**
I was also able to reproduce the same issue with Storybook 6.4.13 and Angular 12.2.15.

Related to #15590 as it shows the same error message and may have the same root cause, however here it can be triggered via the Storybook-Docs Addon without even adding a new Story to the base project template. Note that using the legacy renderer via the parameter `angularLegacyRendering: true` does not solve the issue (stories are only rendered in the docs page in this case, but fail to load with different errors on the standalone story pages).

**Comment by kroeder** [comment]
I can confirm this. With 6.5.0-alpha.21 I run into both of the above-mentioned errors. I believe one causes the other.

I changed the MDX code so that it does not use the embed feature 

```mdx
<Meta title="Documentation/Button" component={ButtonComponent} />

export const Template = (args) => ({ props: args });

<Canvas>
    <Story name="foo">
        {Template.bind({})}
    </Story>
</Canvas>

### Secondary

<Canvas>
    <Story name="bar">
        {Template.bind({})}
    </Story>
</Canvas>
```

Now both errors are gone and I am no longer able to cause these errors by switch back and forth between stories.

# React error

### Warning: render(...): It looks like the React-rendered content of this container was removed without using React.

My first assumption would be that something does not work between the change of the DocsRenderer and CanvasRenderer. It's the same story but if you change back and forth between Canvas and Docs with the same story then the errors appear 

By same story I mean 

**MDX / DocsRenderer**
```
<Canvas>
  <Story id="example-button--primary" />
</Canvas>

<Canvas>
  <Story id="example-button--secondary" />
</Canvas>
```

**TS / CanvasRenderer**
```typescript
export default {
  title: 'Example/Button',
  component: Button,
  // More on argTypes: https://storybook.js.org/docs/angular/api/argtypes
  argTypes: {
    backgroundColor: { control: 'color' },
  },
} as Meta;

// More on component templates: https://storybook.js.org/docs/angular/writing-stories/introduction#using-args
const Template: Story<Button> = (args: Button) => ({
  props: args,
});

export const Primary = Template.bind({});
// More on args: https://storybook.js.org/docs/angular/writing-stories/args
Primary.args = {
  primary: true,
  label: 'Button',
};

export const Secondary = Template.bind({});
Secondary.args = {
  label: 'Button',
};```

# Angular error

I believe the "part of 2 modules" error appears because the react error prevents the code from executing ?`resetCompiledComponents()` in the angular renderer.

**Comment by sir-captainmorgan21** [comment]
I can confirm. I am running into this same issue. When navigating from docs to stories, the story loads and then disappears with the error in the console. I also noticed that it does not appropriately switch to the canvas tab when switching from a docs only page to a story.

**Comment by sir-captainmorgan21** [comment]
@shilman Sorry just wanted to ping you on any possible updates on this? This is a pretty impactful bug for us right now. That being said I know its one of many many issues

**Comment by jimyhdolores** [comment]
Any news with this? I have the same problem

**Comment by vincentB-lemon** [comment]
I have tried multiple versions of addon-docs (6.4.19, 6.4.20, 6.5.0-alpha.59) and with each of them, I still get the same error. On which version did you get it fixed @sir-captainmorgan21 ?

**Comment by sir-captainmorgan21** [comment]
@vincentB-lemon we are 6.4.19

**Comment by DzmVasileusky** [comment]
Same bug for v6.5.5

This simple code is causing it:
```
import { FormsModule } from '@angular/forms';
import { Canvas, Meta, Story } from '@storybook/addon-docs';
import { moduleMetadata } from '@storybook/angular';
import { CuiDatepickerModule } from './datepicker.module';

<Meta
    decorators={[
        moduleMetadata({
            imports: [CuiDatepickerModule, FormsModule]
        })
    ]}
    title="Components/Datepicker"
/>

export const DatepickerTemplate = (args) => ({
    props: args,
    template: `<cui-datepicker [ngModel]="value1"></cui-datepicker>`
});

<Canvas>
    <Story name="Overview">
        {DatepickerTemplate.bind({})}
    </Story>
</Canvas>
```

**Comment by Velmoren** [comment]
Same error in version 6.5.9

Appears after story output by its id

in Properties.stories.mdx

```
<Canvas>
    <Story
        name="Red"
        args={{
            label: 'Red',
            color: 'red'
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>
```

in Overwrite.stories.mdx

```
<Canvas>
    <Story id="components-buttons-overwrites--blue"/>
</Canvas>
```

after changing routes I get an error

```
Unhandled Promise rejection: The selector "sb-components-buttons-overwrites--blue-hnxnwznbq-component" did not match any elements ; Zone: <root> ; Task: Promise.then ; Value: Error: The selector "sb-components-buttons-overwrites--blue-hnxnwznbq-component" did not match any elements
```

**Comment by axeljaeger-arculus** [comment]
I am facing the exact same two issues. However, I was able to solve the first one with the double declaration by putting my component into its own module and import that module in the moduleMetadata-block.

Now my story is working on both Canvas and Docs tab, however, I still get the Unhandled Promise reject as Velmoren. It might have something to do with the mangled-name. I saw previously similar errors but without the mangled name.

**Comment by axeljaeger-arculus** [comment]
Setting `modernInlineRender: true` solved the problem for me as stated here:
https://bytemeta.vip/repo/storybookjs/storybook/issues/18175

**Comment by Velmoren** [comment]
> Setting `modernInlineRender: true` solved the problem for me as stated here: https://bytemeta.vip/repo/storybookjs/storybook/issues/18175

I get this error with `modernInlineRender: true` as well.

`
module.exports = {
    stories: [
        "../src/**/*.stories.mdx",
        "../src/**/*.stories.@(js|jsx|ts|tsx)"
    ],
    addons: [
        "@storybook/addon-links",
        "@storybook/addon-essentials",
        "@storybook/addon-interactions",
        "storybook-addon-themes"
    ],
    framework: "@storybook/angular",
    core: {
        "builder": "@storybook/builder-webpack5"
    },
    modernInlineRender: true
}
`

[https://postimg.cc/94RH4K4G](url)

And when I reload the page, I get No code available

[https://postimg.cc/Yh07Kvpn](url)

The above setting does not help at all. Ready to provide my git with this error. Maybe we just write something wrong

**Comment by valentinpalkovic** [comment]
@Velmoren which storybook Version do you use?

**Comment by Velmoren** [comment]
> @Velmoren Which Storybook version do you use?

A very annoying bug. I can take a video. Basically there are two problems:
1) Doesn't find component selector when changing from MDX to CSF
2) The Show code button is not always active. I noticed that after switching from component to component, it appears, and when the page with CSF is first loaded, the button is inactive.

Storybook v 6.5.16
Angular v 13.3.9

**Comment by valentinpalkovic** [comment]
@Velmoren Would it be an option for you to upgrade to the latest SB 7.0 beta release? https://storybook.js.org/blog/7-0-beta/

You can follow the migration guide if you want: https://chromatic-ui.notion.site/Storybook-7-migration-guide-dbf41fa347304eb2a5e9c69b34503937
We improved the Angular support in SB 7.0 a lot and fixed also some bugs related to Docs and MDX. Especially the mentioned problem of this GH issue should be fixed in SB 7.0 beta.

**Comment by Velmoren** [comment]
@valentinpalkovic 
SB 7.0 beta supports Angular 14+, our design system is based on Angular 13. Switching to Angular 14 is too complicated for our system.
In the framework of 6.5 this problem can not be solved?

**Comment by Velmoren** [comment]
`@valentinpalkovic `
https://youtu.be/31Bb7_rCBNk

**Comment by valentinpalkovic** [comment]
Unfortunately no. At least I am not beware of a quick fix :/. We only port critical hotfixes back to 6.5.

**Comment by Velmoren** [comment]
> Unfortunately no. At least I am not beware of a quick fix :/. We only port critical hotfixes back to 6.5.

Sad, very sad. If there are any changes in this direction, please write in this thread.

**Comment by valentinpalkovic** [comment]
Closing, as this is fixed in the latest SB 7.0 beta.
